libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/snapshotter/api/MapboxSnapshotterApi.kt

```diff
@@ -64,129 +65,171 @@ class MapboxSnapshotterApi(
-                        it.insets,
-                        it.bearing,
-                        it.pitch
-                    )
-                    val centerTop =
-                        (
-                            (mapInterface.size.height * options.density) -
-                                (options.edgeInsets.top + options.edgeInsets.bottom)
-                            ) / 2 + (options.edgeInsets.top)
-                    val centerLeft =
-                        (
-                            (mapInterface.size.width * options.density) -
-                                (options.edgeInsets.left + options.edgeInsets.right)
-                            ) / 2 + (options.edgeInsets.left)
-                    cameraOptions.padding = getEdgeInsets(
-                        Size(
-                            mapInterface.size.width * options.density,
-                            mapInterface.size.height * options.density
-                        ),
-                        ScreenCoordinate(
-                            centerLeft,
-                            centerTop
-                        )
-                    )
-                    snapshotter.setCameraOptions(cameraOptions)
-                    snapshotter.setUri(options.styleUri)
-                    mapInterface.size = oldSize
-                    snapshotter.setStyleListener(object : SnapshotStyleListener {
-                        override fun onDidFinishLoadingStyle(style: Style) {
-                            style.addSource(
-                                geoJsonSource(RouteConstants.PRIMARY_ROUTE_SOURCE_ID) {
-                                    geometry(LineString.fromLngLats(routeLinePoints))
-                                }
-                            )
-                            style.addLayer(
-                                (
-                                    SnapshotterProcessor
-                                        .process(SnapshotterAction.GenerateLineLayer)
-                                        as SnapshotterResult.SnapshotLineLayer
-                                    ).layer
-                            )
-                        }
-                    })
-                    snapshotter.start(object : SnapshotCreatedListener {
-                        override fun onSnapshotResult(snapshot: MapSnapshotInterface?) {
-                            val bitmapAction = SnapshotterAction.GenerateBitmap(
-                                options,
-                                snapshot
-                            )
+    fun generateSnapshot(
+        progress: RouteProgress,
+        consumer: MapboxNavigationConsumer<Expected<SnapshotValue, SnapshotError>>
+    ) {
+        val bannerInstructions = progress.bannerInstructions
+        ifNonNull(bannerInstructions) { instruction ->
+            val action = SnapshotterAction.GenerateSnapshot(instruction)
+            val result = SnapshotterProcessor.process(action)
+            when (result) {
```

Comment:
Could we extract some of the logic into well-named `private` `fun`ctions? `when` logic is getting really long.

Reply:
I already did as much as could be done. I don't think there is any scope to extract it further

Reply:
How about extracting `is SnapshotterResult.SnapshotAvailable` branch code into a `private` `fun`ction?
Also,

```kotlin
consumer.accept(
                                        Expected.Failure(
                                            SnapshotError(
                                                "Inappropriate $result emitted for $action.",
                                                null
                                            )
                                        )
                                    )
```

can be extracted as it's duplicated in `is SnapshotterResult.Snapshot.Failure` and 

```
is SnapshotterResult.SnapshotterCameraPosition,
is SnapshotterResult.Snapshot.Success,
is SnapshotterResult.Snapshot.Failure,
is SnapshotterResult.SnapshotLineLayer
```

Is that an option?

Reply:
I already did in this [PR](https://github.com/mapbox/mapbox-navigation-android/pull/4125#discussion_r592716451) but it wasn't taken well.